### PR TITLE
CORS-3033: Bring your own DNS for Azure

### PR DESCRIPTION
**Pull request by jhixson74**
Bring your own DNS for Azure

[Copied from https://github.com/openshift/installer/pull/7631]

```
New functionality for updating ignition data for the bootstrap node. After the terraform stage that establishes the public and private load balancer ips/dns date is applied:

    The output is read / parsed
    The load balancer data is added to a new file
    The bootstrap ignition file is regenerated
    The contents of the ignition file are used to update the file that contains the terraform variables
```

https://issues.redhat.com/browse/CORS-3033
https://issues.redhat.com/browse/CORS-3034
https://issues.redhat.com/browse/CORS-3035
https://issues.redhat.com/browse/CORS-3036
https://issues.redhat.com/browse/CORS-3037
https://issues.redhat.com/browse/CORS-3038
https://issues.redhat.com/browse/CORS-3039
https://issues.redhat.com/browse/CORS-3040

### CODE REVIEW

File: pkg/types/dns/dns.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+	UserProvisionedDNSDisabled UserProvisionedDNS = "Disabled"
+)
+
+// ValidUserProvisionedDNSType verifies the userProvisionedDNS string is valid
+func ValidUserProvisionedDNSType(userProvisionedDNS UserProvisionedDNS) bool {
+	if userProvisionedDNS == UserProvisionedDNSEnabled || userProvisionedDNS == UserProvisionedDNSDisabled {
+		return true
+	}
+	return false
+}
```
Comment (r4f4) [inline pkg/types/dns/dns.go:20]: This dns package change could be its own commit.

File: pkg/terraform/stages/azure/stages.go [hunk 2]
```diff
@@ -46,3 +55,82 @@ var StackPlatformStages = []terraform.Stage{
 		[]providers.Provider{providers.AzureStack},
 	),
 }
+
+// extractAzureLBConfig extracts the load balancer information from the terraform outputs, generates the
+// Load Balancer Config file, regenerates the bootstrap ignition, and updates the terraform variables file.
+func extractAzureLBConfig(s stages.SplitStage, directory string, terraformDir string, file *asset.File, tfvarsFile *asset.File) (string, error) {
```
Comment (r4f4) [inline pkg/terraform/stages/azure/stages.go:61]: The AWS version of this function looks like an identical copy. Maybe this could be extracted out from the individual providers and a common function could be used instead?

File: pkg/types/dns/dns.go (hunk 1 above)
Comment (jhixson74) [reply pkg/types/dns/dns.go:20]: Agree. I'll separate it.

File: data/data/azure/vnet/main.tf [hunk 3]
```diff
@@ -82,7 +82,7 @@ resource "azurerm_storage_account" "cluster" {
     }
   }
 
-  dynamic identity {
+  dynamic "identity" {
```
[the remark this answers is not in the document]
Comment (r4f4) [reply data/data/azure/vnet/main.tf:85]: Which terraform binary are you using? Unfortunately, the Installer uses an ancient one in CI: https://github.com/openshift/installer/blob/master/hack/tf-fmt.sh#L15

[the remark this answers is not in the document]
Comment (jhixson74) [reply data/data/azure/vnet/main.tf:85]: That needs to be updated ;-) We are currently using v1.3.7 in the installer. Using v0.12.0-rc1 for validation seems silly.

File: pkg/asset/cluster/tfvars.go
Comment (r4f4) [inline pkg/asset/cluster/tfvars.go:1]: Nit: these changes should probably be in the first commit.

File: pkg/types/dns/dns.go [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+	// UserProvisionedDNSDisabled indicates that the DNS solution is provisioned by the Installer.
+	UserProvisionedDNSDisabled UserProvisionedDNS = "Disabled"
+)
+
+// ValidUserProvisionedDNSType verifies the userProvisionedDNS string is valid
+func ValidUserProvisionedDNSType(userProvisionedDNS UserProvisionedDNS) bool {
+	if userProvisionedDNS == UserProvisionedDNSEnabled || userProvisionedDNS == UserProvisionedDNSDisabled {
+		return true
+	}
+	return false
```
Comment (r4f4) [inline pkg/types/dns/dns.go:19]: Nit:
```suggestion
	return userProvisionedDNS == UserProvisionedDNSEnabled || userProvisionedDNS == UserProvisionedDNSDisabled
```

**Review by r4f4 — COMMENTED** [review]
Commit structure is good.

File: pkg/terraform/stages/gcp/stages.go [hunk 5]
```diff
@@ -24,7 +16,7 @@ var PlatformStages = []terraform.Stage{
 		"gcp",
 		"cluster",
 		[]providers.Provider{providers.Google},
-		stages.WithCustomExtractLBConfig(extractGCPLBConfig),
+		stages.WithCustomExtractLBConfig(stages.ExtractGCPLBConfig),
```
Comment (r4f4) [inline pkg/terraform/stages/gcp/stages.go:19]: Shouldn't this be
```suggestion
		stages.WithCustomExtractLBConfig(stages.ExtractLBConfig),
```

File: pkg/types/dns/dns.go (hunk 1 above)
Comment (jhixson74) [reply pkg/types/dns/dns.go:20]: done.

File: pkg/terraform/stages/azure/stages.go (hunk 2 above)
Comment (jhixson74) [reply pkg/terraform/stages/azure/stages.go:61]: done.

File: pkg/types/dns/dns.go (hunk 4 above)
Comment (jhixson74) [reply pkg/types/dns/dns.go]: updated.

**Comment by r4f4** [comment]
Smart move unifying the terraform variables! LGTM but will wait on test results before stamping.

**Comment by r4f4** [comment]
Are the changes in https://github.com/openshift/installer/pull/7775/commits/98e824548c40fd8acdfcd93568c1c2469b02e56c correct? There is a `resource "azurerm_storage_account" "cluster"` being added which I think was moved to a bootstrap stage so the storage account is deleted with bootstrap destroy.

**Comment by jhixson74** [comment]
> > Are the changes in [98e8245](https://github.com/openshift/installer/commit/98e824548c40fd8acdfcd93568c1c2469b02e56c) correct? There is a `resource "azurerm_storage_account" "cluster"` being added which I think was moved to a bootstrap stage so the storage account is deleted with bootstrap destroy.
> 
> I have no idea why that is stuck there. It came in with a rebase and apparently stuck. I will get it out.

I think I know why. The tf-fmt changed it so it thin

> > Are the changes in [98e8245](https://github.com/openshift/installer/commit/98e824548c40fd8acdfcd93568c1c2469b02e56c) correct? There is a `resource "azurerm_storage_account" "cluster"` being added which I think was moved to a bootstrap stage so the storage account is deleted with bootstrap destroy.
> 
> I have no idea why that is stuck there. It came in with a rebase and apparently stuck. I will get it out.

I think the tf-fmt added it since it got modified. Good catch.

**Review by r4f4 — COMMENTED** [review]
/lgtm

**Comment by r4f4** [comment]
Azure job is broken:
```
level=error msg=Error: Null value found in list
level=error
level=error msg=  with module.dns[0].azurerm_private_dns_a_record.apiint_internal[0],
level=error msg=  on dns/dns.tf line 35, in resource "azurerm_private_dns_a_record" "apiint_internal":
level=error msg=  35:   records             = [var.internal_lb_ipaddress_v4]
level=error
level=error msg=Null values are not allowed for this attribute value. 
```
